### PR TITLE
QL: wildcard field type support

### PR DESCRIPTION
**Pull request by astefan**
This PR adds support for `wildcard` data type following its addition in Elasticsearch with https://github.com/elastic/elasticsearch/pull/49993.

There is one issue that will need to be addressed afterwards: there is a slightly different behavior between `wildcard` and the other text based data types (ie `keyword`) when it comes to Painless scripting. As soon as https://github.com/elastic/elasticsearch/issues/58044 is fixed, the workaround in `InternalQlScriptUtils` needs to be reverted.

Addresses https://github.com/elastic/elasticsearch/issues/54184.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/whitelist/InternalQlScriptUtils.java
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/whitelist/InternalQlScriptUtils.java:37]: The issues seems fixed, can you remove the workaround?

File: x-pack/plugin/sql/qa/server/src/main/resources/wildcard.csv-spec
Comment (matriv) [inline x-pack/plugin/sql/qa/server/src/main/resources/wildcard.csv-spec:59]: Wouldn't `groupBy` be more appropriate?

File: x-pack/plugin/sql/qa/server/src/main/resources/wildcard.csv-spec
Comment (matriv) [inline x-pack/plugin/sql/qa/server/src/main/resources/wildcard.csv-spec:76]: and here `complexGroupBy` ?

**Review by matriv — COMMENTED** [review]
Looks good! Left some minor comments.
Also:
- Can we have some Jdbc (PreparedStatement set() and ResultSet get()) tests.
- Is it safe to merge it for 7.9 without the JDBC compatibility tests in place?

**Review by costin — APPROVED** [review]
LGTM. I too am concern about the lack of backwards compatibility...

**Review by bpintea — COMMENTED** [review]
LGTM, except for the compatibility bit.
Generally looking at it, the new type doesn't seem to have an echo in the SQL functionality itself (much like `constant-keyword` vs `keyword`): are the three types distinguishable at SQL level, characteristics- and behavior- or restrictions-wise? B/c if they aren't, while strictly conceptually different types should indeed be reported, I'm wondering if we could compromise to ease the backwards compatibility restrains.

**Comment by astefan** [comment]
Currently blocked by: #56722
We need the bwc tests in place before adding a new data type.

**Comment by costin** [comment]
>  B/c if they aren't, while strictly conceptually different types should indeed be reported, I'm wondering if we could compromise to ease the backwards compatibility restrains.

There is [ongoing work](https://github.com/elastic/elasticsearch/issues/53175#issuecomment-643444444) on this front.

**Comment by astefan** [comment]
When the work in https://github.com/elastic/elasticsearch/issues/53175 is done, this PR is not needed anymore. Also, CONSTANT_KEYWORD data type will also have to be removed from QL.

**Comment by costin** [comment]
See #58315 as well.

**Comment by costin** [comment]
What's the status of this PR considering #58315 and #53175 have been closed?

**Comment by astefan** [comment]
@costin this PR has been updated after those PRs have been merged and now it contains only test code with the `wildcard` field. A `wildcard` field is regarded as a normal `keyword` field.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/Types.java
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/Types.java:52]: Why is this needed? Aren't wildcard returned as keywords? Or is this using a different API/source?

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/Types.java
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/Types.java:52]: That's for testing. We use .json files having an index' complete mapping (see `org.elasticsearch.xpack.ql.type.TypesTests`) and the `wildcard` comparison is for bypassing SQL's and EQL's specific data types auto-resolver in the specific case of `wildcard`.

**Pull request merged**
